### PR TITLE
Blocks indexation #195 

### PR DESCRIPTION
**Pull request by otselnik**
Added a table to store the correspondence between a block hash and a slot, and for constants.
For canceled transactions, the slot is set to the slot of the cancel transaction instead of null.
The block hash table is filled in portions for a given number of transactions periodically after searching for new transactions.
Process blocks with a transaction with results first.
Fill block hash info in transaction logs on processing user requests.
In solana to ethereum transactions table store transaction position in a sequence.

### CODE REVIEW

File: proxy/indexer/solana_receipts_update.py [hunk 1]
```diff
@@ -37,26 +41,30 @@ def __init__(self, signature, results):
 class Indexer:
     def __init__(self):
         self.client = Client(solana_url)
+        self.blocks_by_hash = SqliteDict(filename="local.db", tablename="solana_blocks_by_hash", autocommit=True)
         self.transaction_receipts = SqliteDict(filename="local.db", tablename="known_transactions", autocommit=True, encode=json.dumps, decode=json.loads)
         self.ethereum_trx = SqliteDict(filename="local.db", tablename="ethereum_transactions", autocommit=True, encode=json.dumps, decode=json.loads)
         self.eth_sol_trx = SqliteDict(filename="local.db", tablename="ethereum_solana_transactions", autocommit=True, encode=json.dumps, decode=json.loads)
-        self.sol_eth_trx = SqliteDict(filename="local.db", tablename="solana_ethereum_transactions", autocommit=True)
+        self.sol_eth_trx = SqliteDict(filename="local.db", tablename="solana_ethereum_transactions", autocommit=True, encode=json.dumps, decode=json.loads)
```
Comment (s-medvedev) [inline proxy/indexer/solana_receipts_update.py:48]: Why we need both `ethereum_solana_tranasctions` and `solana_ethereum_transactions` tables?

File: proxy/plugin/solana_rest_api.py [hunk 2]
```diff
@@ -136,39 +139,78 @@ def eth_getBalance(self, account, tag):
...
+                            trx['transactionIndex'] = hex(trx_index)
+                            trx_index += 1
+                            transactions.append(trx)
+                    else:
+                        transactions.append(eth_trx['eth'])
+
+        ret = {
+            "gasUsed": hex(gasUsed),
+            "hash": '0x' + base58.b58decode(block_info['blockhash']).hex(),
+            "number": hex(block_info['blockHeight']),
```
Comment (s-medvedev) [inline proxy/plugin/solana_rest_api.py]: Is it possible that the block number in the response will not coincide with the number transferred to the `eth_getBlockByNumber` request?

File: proxy/plugin/solana_rest_api.py [hunk 3]
```diff
@@ -136,39 +139,78 @@ def eth_getBalance(self, account, tag):
...
+            "parentHash": '0x' + base58.b58decode(block_info['previousBlockhash']).hex(),
+            "timestamp": hex(block_info['blockTime']),
+            "transactions": transactions,
+            "logsBloom": '0x'+'0'*512,
```
Comment (s-medvedev) [inline proxy/plugin/solana_rest_api.py:176]: What information should be passed in `logsBloom` field? I think we need to fill it. Create a separate issue for that.

File: proxy/test_eth_sendRawTransaction.py [hunk 4]
```diff
@@ -139,6 +144,18 @@ def deploy_test_185_solidity_contract(self):
             abi=test_185_solidity_contract.abi
         )
 
+    # @unittest.skip("a.i.")
+    def test_check_get_block_by_hash(self):
+        print("\ntest_check_get_block_by_hash")
+        block = proxy.eth.get_block(self.deploy_block_hash, full_transactions=True)
```
Comment (s-medvedev) [inline proxy/test_eth_sendRawTransaction.py:150]: Please, use assertEqual to check data correctness.

File: proxy/plugin/solana_rest_api.py (hunk 2 above)
Comment (otselnik) [reply proxy/plugin/solana_rest_api.py]: Forgot about this line when move from height to slots. Fixed.

File: proxy/indexer/solana_receipts_update.py (hunk 1 above)
Comment (otselnik) [reply proxy/indexer/solana_receipts_update.py:48]: SqliteDict allow search only by key.

File: proxy/test_eth_sendRawTransaction.py (hunk 4 above)
Comment (otselnik) [reply proxy/test_eth_sendRawTransaction.py:150]: I added tests to `getLogs` pull request. There is more work done to successfully pass tests. 

https://github.com/neonlabsorg/proxy-model.py/pull/214/commits/6c02f548f2d4477eec7a0488e2212e787b1e5fcf

File: proxy/plugin/solana_rest_api.py (hunk 3 above)
Comment (otselnik) [reply proxy/plugin/solana_rest_api.py:176]: Created issue https://github.com/neonlabsorg/proxy-model.py/issues/216

**Review by s-medvedev — APPROVED** [review]

**Pull request merged**